src/tools/w3dviewer/graphicview.cpp

```diff
@@ -0,0 +1,1095 @@
+            m[1].W = 0.0f;
+            m[2].X = 0.0f;
+            m[2].Y = 1.0f;
+            m[2].Z = 0.0f;
+            m[2].W = 0.0f;
+            tm = tm * m;
+        }
+
+        m_camera->Set_Transform(tm);
+    }
+
+    CMainFrame *frame = reinterpret_cast<CMainFrame *>(AfxGetMainWnd());
+
+    if (frame != nullptr) {
+        frame->UpdatePolyCount(robj->Get_Num_Polys());
+    }
+}
+
+void CGraphicView::ResetCameraValues(SphereClass &sphere)
+{
+    m_radius = sphere.Radius * 3.0f;
+
+    if (m_radius < 1.0f) {
+        m_radius = 1.0f;
+    }
+
+    Matrix3D tm;
+    tm.Look_At(Vector3(m_radius + sphere.Center.X, sphere.Center.Y, sphere.Center.Z), sphere.Center, 0.0f);
+    Pos = sphere.Center;
+    m_objCenter = Pos;
+    Zoom = m_radius * 0.0052631581f;
+    Rotation = Build_Quaternion(tm);
+    m_camera->Set_Transform(tm);
+    CW3DViewDoc *document = reinterpret_cast<CW3DViewDoc *>(m_pDocument);
+    LightClass *light = document->m_light;
+
+    if (m_light != nullptr && light != nullptr) {
+        tm.Make_Identity();
+        tm.Adjust_Translation(Pos);
+        tm[2].W += m_radius * 0.69999999f;
+        light->Set_Transform(tm);
+        m_light->Set_Transform(tm);
+        m_light->Scale(m_radius / Lightradius * 14.0f);
+        Lightradius = m_radius * 0.071428575f;
+    }
+
+    float zfar = m_radius * 60.0f;
+    float znear = Zoom * 0.5f;
+
+    if (znear < 0.2f) {
+        znear = 0.2f;
+    }
+
+    if (!document->m_useManualClipPlanes) {
+        m_camera->Set_Clip_Planes(znear, zfar);
+        float start;
+        float end;
+        document->m_scene->Get_Fog_Range(&start, &end);
+        document->m_scene->Set_Fog_Range(end, end);
+        document->m_scene->Update_Fog_Range();
+    }
+
+    document->m_docCamera->Set_Transform(tm);
+    document->m_docCamera->Set_Position(Vector3(0.0f, 0.0f, 0.0f));
+
+    CMainFrame *frame = reinterpret_cast<CMainFrame *>(AfxGetMainWnd());
+
+    if (frame != nullptr) {
+        frame->UpdateCameraDistance(m_radius);
+        frame->UpdateFrameCount(0, 0, 0.0f);
+    }
+
+    m_objSphere = sphere;
+}
+
+void CGraphicView::ResetParticleEmitterCamera(ParticleEmitterClass *emitter)
+{
+    ParticleBufferClass *buffer = emitter->Peek_Buffer();
+    SphereClass s;
+    s.Center = emitter->Get_Start_Velocity();
+    Vector3 acceleration = buffer->Get_Acceleration();
+    float lifetime = buffer->Get_Lifetime();
+    float f1 = lifetime * lifetime;
+    float f2 = acceleration.X * f1;
+    float f3 = acceleration.Y * f1;
+    float f4 = f2 * 0.5f;
+    float f5 = f3 * 0.5f;
+    float f6 = s.Center.Y * lifetime;
+    float f7 = s.Center.Z * lifetime;
+    float f8 = s.Center.X * lifetime + f4;
+    float f9 = 0.0f;
+    float f10 = 0.0f;
+    float f11 = f6 + f5;
+    float f12 = 0.0f;
+    float f13 = f1 * acceleration.Z * 0.5f + f7;
+    float f14;
+
+    if (acceleration.X == 0.0f) {
+        if (acceleration.Y == 0.0f && acceleration.Z == 0.0f) {
+            goto l1;

```

Comment:
You can rewrite this condition to get rid of the `goto`.

```c++
if (acceleration.X != 0.0f && acceleration.Y != 0.0f && acceleration.Z != 0.0f) {
    if (acceleration.X == 0.0f) {
        f14 = 0.0f;
    } else {
        f14 = -s.Center.X / acceleration.X;
    }
    // ... all the other code until the goto label
}
```

Reply:
fixed.